### PR TITLE
Fix AccessTokenUtil does not url encode its parameters

### LINKED ISSUE #40616: Keyvault JCA's AccessTokenUtil does not urlencode its parameters when getting an access token
**Describe the bug**
In our app, we get `NullPointerException` when first accessing  a key through a `com.azure.security.keyvault.jca.implementation.KeyVaultClient` method.

We have all the reasons to believe our configuration is OK, as multiple other KeyVault-based elements in the stack work OK (e.g. KeyVault as property source, SecretClient instances...). The configuration / interaction mode with keyvault is client_id / client_secret based.

On top of it, the same code works on many environments (preproduction...) but the NPE occurs only in production, with the seemingly only relevant difference being the actual values client_id / client_secert.

Finally, it's pretty clear to me, having looked at it for a few hours that the issue is : 
a) hidden by the fact that the HTTP interactions made by the library output nothing / no log / error when HTTP error occurs (40x, 50x, ...)
b) triggered while accessing the accessToken by tenant id / client id / client secret. 
c) more excplicetly, because the `client_secret` is not URL-encoded, which triggers a (rightfull) signin failed error on  `https://login.microsoftonline.com`'s side : the secret is "wrong"

***Exception or Stack Trace***
```
Factory method 'myBeanMethod' 
  threw exception; nested exception is java.lang.NullPointerException: 
  Cannot invoke "com.azure.security.keyvault.jca.implementation.model.AccessToken.getAccessToken()" 
    because "this.accessToken" is null
```

Note : there is no stack trace for the HTTP-call that tries to retrieve the token but Azure Portal (Enterprise application sign-in logs) show a failed login attempt, in our case, with an  `Invalid client secret` failure reason.

**To Reproduce**
- Create an azure entreprise application with a client secret that requires escaping to conform to `x-www-form-urlencoding` (eg `@` or `[]` come to mind
- Instantiate a `com.azure.security.keyvault.jca.implementation.KeyVaultClient` using the valid clientid / secret / tenant ID / keyvault URL
- Perform an operation (getKey, getCertificate)
- NPE is thrown

***Code Snippet***
```
	private void azureJcaKeyVaultClient(AzureKeyVaultSecretProperties kvProperties) {
		String clientId = kvProperties.getCredential().getClientId();
		String clientSecret = kvProperties.getCredential().getClientSecret();
		String tenantId = kvProperties.getProfile().getTenantId();
		String endpoint = kvProperties.getEndpoint();

		var c = new KeyVaultClient(endpoint, tenantId, clientId, clientSecret);
                c.getKey("someKey");
	}

```

**Expected behavior**
Have a successfull login when the client secret (or any other authentication variable) require encoding to conform to the HTTP status

**Screenshots**
N/A

**Setup (please complete the following information):**
 - OS: Linux, MacOS
 - IDE: IntelliJ, deployed on Azure Linux ASP
 - Library/Libraries: com.azure:azure-security-keyvault-jca:2.8.1
 - Java version: 11, 17, 21
 - App Server/Environment: SpringBoot / Tomcat

**Additional context**
The bug is pretty clear when looking at current implementation : 

https://github.com/Azure/azure-sdk-for-java/blob/dcbfa344c9dc0696fac8f4da29d6ef8ac9085c9b/sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java#L93-L111

Raw string concatenation may or may not work considering the request's content type.

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X] Bug Description Added
- [X] Repro Steps Added
- [X] Setup information Added

### PR DESCRIPTION
**Pull request by rujche**
# Description

Fix #40616

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java [hunk 1]
```diff
@@ -99,9 +102,9 @@ public static AccessToken getAccessToken(String resource, String aadAuthenticati
 
         StringBuilder requestBody = new StringBuilder();
         requestBody.append(GRANT_TYPE_FRAGMENT)
-                   .append(CLIENT_ID_FRAGMENT).append(clientId)
-                   .append(CLIENT_SECRET_FRAGMENT).append(clientSecret)
-                   .append(RESOURCE_FRAGMENT).append(resource);
+            .append(CLIENT_ID_FRAGMENT).append(URLEncoder.encode(clientId, StandardCharsets.UTF_8))
+            .append(CLIENT_SECRET_FRAGMENT).append(URLEncoder.encode(clientSecret, StandardCharsets.UTF_8))
+            .append(RESOURCE_FRAGMENT).append(URLEncoder.encode(resource, StandardCharsets.UTF_8));
```
Comment (rujche) [inline sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java:107]: FYI: 
1. According to the ms doc, every parameters should url encoded, not encode one specific parameter nor the whole requestBody.
2. According to the ms doc, the parameter should be `scope`, not `resource`. I don't know why it use `resource` here. Anyway, I just keep the original behavior.

Refs: https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#first-case-access-token-request-with-a-shared-secret

**Comment by rujche** [comment]
Hi, @saragluna , please help to review this PR.

**Review by saragluna — APPROVED** [review]
Is this change tested?

**Comment by rujche** [comment]
> Is this change tested?

@saragluna 
Tested in my localhost.
Seems `java -keyvault - ci` skipped this test. I'll try to use `java - keyvault - tests` to test it.

**Pull request merged**